src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -912,23 +904,17 @@ private AttributeListSyntax ParseAttributeDeclaration()
             }
 
             var attributes = _pool.AllocateSeparated<AttributeSyntax>();
-            try
+            if (attrLocation != null && attrLocation.Identifier.ToAttributeLocation() == AttributeLocation.Module)
             {
-                if (attrLocation != null && attrLocation.Identifier.ToAttributeLocation() == AttributeLocation.Module)
-                {
-                    attrLocation = CheckFeatureAvailability(attrLocation, MessageID.IDS_FeatureModuleAttrLoc);
-                }
-
-                this.ParseAttributes(attributes);
-                var closeBracket = this.EatToken(SyntaxKind.CloseBracketToken);
-                var declaration = _syntaxFactory.AttributeList(openBracket, attrLocation, attributes, closeBracket);
-
-                return declaration;
-            }
-            finally
-            {
-                _pool.Free(attributes);
+                attrLocation = CheckFeatureAvailability(attrLocation, MessageID.IDS_FeatureModuleAttrLoc);
             }
+
+            this.ParseAttributes(attributes);
+            return _syntaxFactory.AttributeList(
+                openBracket,
+                attrLocation,
+                _pool.ToListAndFree(attributes),
+                this.EatToken(SyntaxKind.CloseBracketToken));
```

Comment:
as much as possible, parsing and creation of hte nodes tries to follow the design of a simple recursive descent parser.  e.g. a construct like `X -> A B C ` should be parsed as `return new X(ParseA(), ParseB(), ParseC())`.  This makes the tree structure and recursive nature readily apparent, and gives no opportunity for confusing state tracking or the possibility for other complex code to run in between making htings more difficult to understand.

Note: there is more we can do here.  For example, our list-parsing today always requires statement forms to get the list.  This can be improved so even with complex syntactic constructs, we can model things as simple recursive descent for all portions of hte node.  Leaving that for the future though as that will be a much larger change.